### PR TITLE
Fix compressed read retries in compressed blobs.

### PR DESCRIPTION
**Pull request by rubensf**
The current configuration will try to use the offset of the _compressed_
blob as the new offset for retrying reads. However, the compressed RE
API wants the offset to refer to the _uncompressed_ blobs. Since the
compressed offset will usually be a lower number than the uncompressed
offset, we generally end up with a bigger blob than expected.

Since this Read interruptions are mostly transient (eg context deadlines
exceeded), our higher level retries have been mostly hiding the issue.

Also took the chance to beef up some comments around the thread tricky
compression code.

I'll _eventually_ add retries tests to compressed blobs writes retry tests :)

### CODE REVIEW

**Comment by rubensf** [comment]
In the process, I've removed the retrier based on a wrong digest, since this pretty much should have been our flake reason...

Wdyt?

File: go/pkg/client/retries_test.go [hunk 1]
```diff
@@ -276,58 +288,71 @@ func TestWriteRetries(t *testing.T) {
...
-			blob := []byte("blob")
-			got, err := f.client.ReadBlob(f.ctx, digest.NewFromBlob(blob))
-			if err != nil {
-				t.Errorf("client.ReadBlob(ctx, digest) gave error %s, want nil", err)
-			}
-			if diff := cmp.Diff(blob, got, cmpopts.EquateEmpty()); diff != "" {
-				t.Errorf("client.ReadBlob(ctx, digest) gave diff (-want, +got):\n%s", diff)
-			}
-		})
+		for _, comp := range []bool{false, true} {
```
Comment (ola-rozenfeld) [inline go/pkg/client/retries_test.go:291]: How much longer did it make the test? If it added a few seconds, then suggestion: make compression retry test a separate Test function from this one that includes the sleeping, because these two are really orthogonal. Ditto for files.

File: go/pkg/client/cas.go [hunk 2]
```diff
@@ -938,40 +961,53 @@ func (c *Client) readBlobStreamed(ctx context.Context, d digest.Digest, offset,
 	if limit > 0 && limit < sz {
 		sz = limit
 	}
+	if err = w.Reset(); err != nil {
```
Comment (ola-rozenfeld) [inline go/pkg/client/cas.go:964]: Wait, if you moved this Reset outside of the closure, that tells me you no longer need it to be resettable. And you're removing the functionality of retrying on failed digest verification, too.
So, suggestion: base this PR on a revert of my https://github.com/bazelbuild/remote-apis-sdks/pull/271. Should be much simpler, imo.

File: go/pkg/client/cas.go [hunk 3]
```diff
@@ -843,24 +846,36 @@ func NewCompressedWriteBuffer(w io.WriteCloser) (io.WriteCloser, chan error, err
...
+		// separate thread. As such, we also need a way to signal the main
+		// thread that the decoding has finished - which will have some delay
+		// from the last Write call.
 		_, err := decoder.WriteTo(w)
 		if err != nil {
 			// Because WriteTo returned early, the pipe writers still
 			// have to go somewhere or they'll block execution.
 			io.Copy(ioutil.Discard, r)
 		}
-		w.Close()
```
[the remark this answers is not in the document]
Comment (rubensf) [reply go/pkg/client/cas.go:852]: Under `if d.Size == sz {` in the new file... `if err := wt.Close()`. 

And I thought the CI was doing `features=race` already, _sighs_. Caught a big, thanks for reminding me!

And i don't think we have windows CI.

File: go/pkg/client/retries_test.go (hunk 1 above)
Comment (rubensf) [reply go/pkg/client/retries_test.go:291]: The non race tests still take < 10s (less than 1s of noticeable increase).

I'm not sure they're orthogonal - I'd argue they _should_ be the same. As much as possible, comp vs noncomp should behave the same way.

File: go/pkg/client/cas.go (hunk 2 above)
Comment (ola-rozenfeld) [reply go/pkg/client/cas.go:970]: We'll have to clean this up in the fixit next week, then. There's a lot of stuff here now that no longer should be that will just be confusing. We no longer need the whole ResettableWriter struct and interface, and the refactoring I did that made everything more complicated...
But I'm okay with leaving it as a cleanup task for the next CL.

File: go/pkg/client/retries_test.go (hunk 1 above)
Comment (ola-rozenfeld) [reply go/pkg/client/retries_test.go:291]: Sure, if it doesn't make it noticeably longer, this is fine!

**Review by ola-rozenfeld — APPROVED** [review]
